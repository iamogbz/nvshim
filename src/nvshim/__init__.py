"""Current package version"""
__version__ = "0.0.0"
