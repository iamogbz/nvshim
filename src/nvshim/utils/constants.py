"""nvshim constants"""
from enum import IntEnum


class ErrorCode(IntEnum):
    """nvshim process exit error codes"""

    ENV_NVM_DIR_MISSING = 1003
    EXECUTABLE_NOT_FOUND = 1002
    KEYBOARD_INTERRUPT = 130
    VERSION_NOT_INSTALLED = 1001


shims = {"node", "npm", "npx"}
