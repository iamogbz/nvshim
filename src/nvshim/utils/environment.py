"""Utility constants and functions for environment management"""
import json
import os
from contextlib import contextmanager
from enum import Enum
from typing import Dict

EnvDict = Dict[str, str]


class EnvironmentVariable(Enum):
    """Environment variables nvshim cares about"""

    AUTO_INSTALL = "NVSHIM_AUTO_INSTALL"
    NVM_DIR = "NVM_DIR"
    VERBOSE = "NVSHIM_VERBOSE"


class MissingEnvironmentVariableError(Exception):
    """Error for missing environment variable"""

    def __init__(self, env_var: EnvironmentVariable):
        super().__init__()
        self.env_var = env_var


def _get_env_var(env_var: EnvironmentVariable, raise_missing: bool = False) -> object:
    try:
        return json.loads(os.environ[env_var.value])
    except KeyError as exc:
        if raise_missing:
            raise MissingEnvironmentVariableError(env_var) from exc
        return None
    except json.decoder.JSONDecodeError:
        return os.environ.get(env_var.value)


def _set_envs(values: EnvDict):
    os.environ.clear()
    os.environ.update(**values)


@contextmanager
def process_env(env_vars: EnvDict):
    """Run code with specific enviroment variables that are reset afterwards"""
    prev_env_vars = {**os.environ}
    _set_envs(env_vars)
    yield
    _set_envs(prev_env_vars)


def is_version_auto_install_enabled() -> bool:
    """Return if the auto install environment variable is true or false"""
    return bool(_get_env_var(EnvironmentVariable.AUTO_INSTALL))


def is_verbose_logging() -> bool:
    """Return if verbosity is set using the nvshim environment variable"""
    return bool(_get_env_var(EnvironmentVariable.VERBOSE))


def get_nvm_dir() -> str:
    """Return the path set from the $NVM_DIR environment variable"""
    return str(_get_env_var(EnvironmentVariable.NVM_DIR, True))
