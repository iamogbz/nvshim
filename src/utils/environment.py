from enum import Enum
import json
import os


class EnvironmentVariable(Enum):
    AUTO_INSTALL = "NVSHIM_AUTO_INSTALL"
    NVM_DIR = "NVM_DIR"
    VERBOSE = "NVSHIM_VERBOSE"


class MissingEnvironmentVariableError(Exception):
    def __init__(self, env_var: EnvironmentVariable):
        self.env_var = env_var


def _get_env_var(env_var: EnvironmentVariable, raise_missing: bool = False) -> object:
    try:
        return json.loads(os.environ[env_var.value])
    except KeyError:
        if raise_missing:
            raise MissingEnvironmentVariableError(env_var)
    except json.decoder.JSONDecodeError:
        return os.environ.get(env_var.value)


def is_version_auto_install_enabled() -> bool:
    return bool(_get_env_var(EnvironmentVariable.AUTO_INSTALL))


def is_verbose_logging() -> bool:
    return _get_env_var(EnvironmentVariable.VERBOSE)


def get_nvm_dir() -> str:
    return _get_env_var(EnvironmentVariable.NVM_DIR, True)
