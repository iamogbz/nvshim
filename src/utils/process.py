import os
import subprocess
import sys
from typing import Dict


def _with_venv(env: Dict[str, str]):
    path_key = "PATH"
    env_path = env.get(path_key, "")
    return dict(env, **{path_key: f"venv/bin/:{env_path}"})


def run(*args, **kwargs):
    try:
        subprocess.run(
            args, **kwargs, check=True, env=_with_venv(kwargs.get("env", os.environ))
        )
    except subprocess.CalledProcessError as error:
        sys.exit(error.returncode)
