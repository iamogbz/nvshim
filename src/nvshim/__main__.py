import argparse
import functools
import os
import re
import sys
from typing import Dict, List, Sequence

import semver

from compiler import __version__
from utils import environment, message, process
from utils.constants import ErrorCode


AliasMapping = VersionMapping = Dict[str, str]


class HashableDict(dict):
    def __hash__(self):
        return hash(frozenset(self.items()))


def get_files(path: str) -> [str]:
    """
    Generate the file paths to traverse, or a single path if a file name was given
    """
    if os.path.isdir(path):
        for root, _, files in os.walk(path):
            for name in files:
                yield os.path.join(root, name)
    elif os.path.exists(path):
        yield path


def _run_nvm_cmd(nvm_sh_path: str, args: str) -> process.subprocess.CompletedProcess:
    return process.run(
        f". {nvm_sh_path} && nvm {args}",
        shell="bash",
        encoding="UTF-8",
        capture_output=True,
    )


@functools.lru_cache(maxsize=None)
def get_nvm_stable_version(nvm_dir) -> str:
    output = _run_nvm_cmd(get_nvmsh_path(nvm_dir), "alias stable").stdout
    result = re.sub(r"\x1B[@-_][0-?]*[ -/]*[@-~]", "", str(output).strip())
    return re.findall(r"> v([\w\.]+)\)", result)[0]


def get_nvm_aliases_dir(nvm_dir: str) -> str:
    return os.path.join(nvm_dir, "alias")


def get_nvm_aliases(nvm_dir: str) -> AliasMapping:
    aliases_to_version = HashableDict(
        default="stable", node="stable", stable=get_nvm_stable_version(nvm_dir)
    )
    nvm_aliases_dir = get_nvm_aliases_dir(nvm_dir)
    for file_path in get_files(nvm_aliases_dir):
        rel_path = os.path.relpath(file_path, nvm_aliases_dir)
        with open(file_path) as f:
            aliases_to_version[rel_path] = f.readline().strip()

    return aliases_to_version


def parse_version(version: str) -> semver.VersionInfo:
    try:
        return semver.parse_version_info(
            version[1:] if version.startswith("v") else version
        )
    except ValueError:
        pass


@functools.lru_cache(maxsize=None)
def resolve_alias(
    name: str, alias_mapping: AliasMapping
) -> (semver.VersionInfo, str, [str]):
    """
    Resolve an alias to a semantic version going through multiple mappings
    
    :param name: name of the version alias to resolve
    :type alias_mapping: mapping of alias to version
    :return: (semantic version info, alias name, list of aliases traversed)
    """
    seen_in_order = []
    seen = set()
    while name in alias_mapping and name not in seen:
        name = alias_mapping.get(name)
        seen_in_order.append(name)
        seen.add(name)

    return parse_version(name), name, seen_in_order


@functools.lru_cache(maxsize=None)
def resolve_nvm_aliases(nvm_aliases: AliasMapping) -> AliasMapping:
    """
    Resolve all aliases in a mapping to their version info
    """
    resolved_aliases = {}
    for alias, version in nvm_aliases.items():
        version_info = parse_version(version) or resolve_alias(version, nvm_aliases)[0]

        if version_info:
            resolved_aliases[alias] = str(version_info)

    return resolved_aliases


def get_node_versions_dir(nvm_dir: str) -> str:
    return os.path.join(nvm_dir, "versions", "node")


def get_node_version_bin_dir(node_versions_dir: str, version: str) -> str:
    return os.path.join(node_versions_dir, f"v{version}", "bin")


def get_node_versions(node_versions_dir: str) -> VersionMapping:
    files = os.listdir(node_versions_dir) if os.path.exists(node_versions_dir) else []
    return {
        str(v): get_node_version_bin_dir(node_versions_dir, str(v))
        for v in map(parse_version, files)
    }


def merge_nvm_aliases_with_node_versions(
    nvm_aliases: AliasMapping, node_versions: VersionMapping
) -> VersionMapping:
    alias_versions = {
        alias: node_versions[version]
        for alias, version in nvm_aliases.items()
        if version in node_versions
    }
    return dict(alias_versions, **node_versions)


def get_nvmrc_path(exec_dir: str = os.getcwd()) -> str:
    root_dir = os.path.abspath(os.sep)
    current_dir = exec_dir
    config_file = ".nvmrc"
    config_found = False
    while True:
        current_config = os.path.join(current_dir, config_file)
        config_found = os.path.exists(current_config)
        if config_found or current_dir == root_dir:
            break
        current_dir = os.path.realpath(os.path.join(current_dir, "../"))

    return current_config if config_found else None


def get_nvmrc(nvmrc_path: str = None) -> (str, bool):
    version = "default"
    if nvmrc_path:
        with open(nvmrc_path) as f:
            rc_version = f.readline().strip()
            parsed_version = parse_version(rc_version)
            if parsed_version:
                return str(parsed_version), True
            version = rc_version

    return version, False


def get_nvmsh_path(nvm_dir: str) -> str:
    return os.path.join(nvm_dir, "nvm.sh")


def _pretty_version(version: str, parsed: bool) -> str:
    return f"v{version}" if parsed else version


def get_bin_path(
    *,
    version: str,
    is_version_parsed: bool,
    nvm_aliases: AliasMapping,
    node_versions: VersionMapping,
    bin_file: str,
    node_versions_dir: str,
    nvm_sh_path: str,
):
    version_mapping = merge_nvm_aliases_with_node_versions(nvm_aliases, node_versions)
    try:
        node_path = version_mapping[version]
    except KeyError:
        if not environment.is_version_auto_install_enabled():
            message.print_version_not_installed(
                _pretty_version(version, is_version_parsed)
            )
            sys.exit(ErrorCode.VERSION_NOT_INSTALLED)

        _run_nvm_cmd(nvm_sh_path, f"install {version}")
        node_path = get_node_version_bin_dir(
            node_versions_dir, version=nvm_aliases.get(version, version)
        )

    bin_path = os.path.join(node_path, bin_file)
    if not os.path.exists(bin_path):
        message.print_node_bin_file_does_not_exist(bin_path)
        sys.exit(ErrorCode.EXECUTABLE_NOT_FOUND)

    return bin_path


def parse_args(args: Sequence[str]) -> (argparse.Namespace, List[str]):
    """
    Get the arguments to be used to execute the node binary
    :return: parsed and unknown arguments
    """
    parser = argparse.ArgumentParser(
        description="Launch executable using project or default node version",
        add_help=False,
    )
    parser.add_argument(
        "bin_file",
        metavar="nvx",
        type=str,
        help="filename of the node executable binary",
    )
    parser.add_argument(
        "bin_args",
        metavar="args",
        nargs="*",
        help="arguments passed to the node executable",
    )

    return parser.parse_known_args(args)


def main(version_number: str = __version__):
    message.print_running_version(version_number)
    parsed_args, unknown_args = parse_args(sys.argv[1:])
    nvmrc_path = get_nvmrc_path()
    version, parsed = get_nvmrc(nvmrc_path)
    try:
        nvm_dir = environment.get_nvm_dir()
    except environment.MissingEnvironmentVariableError as error:
        message.print_env_var_missing(error.env_var)
        sys.exit(ErrorCode.ENV_NVM_DIR_MISSING)

    bin_path = get_bin_path(
        version=version,
        is_version_parsed=parsed,
        nvm_aliases=resolve_nvm_aliases(get_nvm_aliases(nvm_dir)),
        node_versions=get_node_versions(get_node_versions_dir(nvm_dir)),
        node_versions_dir=get_node_versions_dir(nvm_dir),
        bin_file=parsed_args.bin_file,
        nvm_sh_path=get_nvmsh_path(nvm_dir),
    )
    message.print_using_version(_pretty_version(version, parsed), bin_path, nvmrc_path)
    process.run(bin_path, *parsed_args.bin_args, *unknown_args)


if __name__ == "__main__":
    main()
