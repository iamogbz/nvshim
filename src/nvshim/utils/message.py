from enum import Enum, IntEnum
from subprocess import CalledProcessError

from colored import fg, stylize

from .environment import EnvironmentVariable


class Color(Enum):
    ERROR = fg("red")
    NOTICE = fg("yellow")


class MessageLevel(IntEnum):
    LOUD = 2
    NORMAL = 1
    QUIET = 0


def _level() -> MessageLevel:
    """Minium threshold level for logging to occur"""
    from .environment import is_verbose_logging

    return MessageLevel.QUIET if is_verbose_logging() else MessageLevel.NORMAL


def _print(*args, level=MessageLevel.NORMAL):
    if level < _level():
        return

    print(*args)


def _stylize(text: str, color: Color) -> str:
    return stylize(text, color.value)


def _print_stylized(text: str, color: Color, level=MessageLevel.NORMAL):
    _print(_stylize(text, color), level=level)


def _print_error(text: str):
    _print_stylized(text, Color.ERROR, MessageLevel.LOUD)


def print_env_var_missing(env_var: EnvironmentVariable):
    _print_error(f"Environment variable '{env_var.value}' missing")


def print_using_version(version: str, bin_path: str, nvmrc_path: str = None):
    messages = (
        f"Found '{nvmrc_path}' with version <{version}>"
        if nvmrc_path
        else f"Using <{version}> version",
        f"\n.{bin_path}\n",
    )
    _print("".join(messages), level=MessageLevel.QUIET)


def print_node_bin_file_does_not_exist(bin_path: str):
    _print(f"No executable file found at '{bin_path}'", level=MessageLevel.LOUD)


def print_version_not_installed(version: str):
    _print_error(f"N/A version '{version}' is not yet installed.\n")
    _print(
        "You need to run",
        _stylize(f"'nvm install {version}'", Color.NOTICE),
        "to install it before using it.\n",
    )
    _print(
        "Or set the environment variable",
        _stylize(f"'{EnvironmentVariable.AUTO_INSTALL.value}'", Color.NOTICE),
        "to auto install at run time.\n",
    )


def print_running_version(version_number: str):
    _print(f"Executing shim version {version_number}", level=MessageLevel.QUIET)


def print_unable_to_get_stable_version(exc: Exception):
    _print_error("Unable to retrieve stable version from nvm")
    _print(str(exc), level=MessageLevel.QUIET)


def print_process_interrupted(exc: KeyboardInterrupt):
    _print(f"\nInterrupted. {exc}")


def print_unable_to_run(exc: CalledProcessError):
    _print(str(exc), level=MessageLevel.QUIET)


def print_unable_to_remove_nvm_shim_temp_file(exc: Exception):
    _print_error("Unable to remove temporary nvm shim file")
    _print(str(exc), level=MessageLevel.QUIET)
