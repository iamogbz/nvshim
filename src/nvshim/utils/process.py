import os
import re
import subprocess
import sys
from typing import Dict

from .constants import ErrorCode
from .environment import EnvironmentVariable, EnvDict, process_env
from .message import print_process_interrupted, print_unable_to_run


def _include_venv(env: EnvDict):
    path_key = "PATH"
    env_path = env.get(path_key, "")
    return {**env, path_key: f"venv/bin/:{env_path}"}


def run(*args, **kwargs) -> subprocess.CompletedProcess:
    env_vars = _include_venv(os.environ)
    env_vars[EnvironmentVariable.AUTO_INSTALL.value] = "false"

    with process_env(env_vars):
        return run_with_error_handler(*args, **kwargs)


def run_with_error_handler(*args, **kwargs) -> subprocess.CompletedProcess:
    try:
        return subprocess.run(args, encoding="UTF-8", **kwargs, check=True)
    except KeyboardInterrupt as interrupt_e:
        print_process_interrupted(interrupt_e)
        sys.exit(ErrorCode.KEYBOARD_INTERRUPT)
    except subprocess.CalledProcessError as process_e:
        print_unable_to_run(process_e)
        sys.exit(process_e.returncode)


def clean_output(output: str) -> str:
    """Removes ansi color codes from string"""
    return re.sub(r"\x1B[@-_][0-?]*[ -/]*[@-~]", "", str(output).strip())
