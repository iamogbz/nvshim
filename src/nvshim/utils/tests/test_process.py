import subprocess
import pytest

from nvshim.utils import constants, process


def test_process_run_completes_successfully():
    output = process.run("echo", "success", stdout=subprocess.PIPE).stdout.strip()
    assert output == "success"


def test_process_run_handles_exception_system_exit():
    with pytest.raises(SystemExit) as exc_info:
        process.run("bash", "-c", "exit 1")

    assert exc_info.value.code == 1


def test_process_run_handles_exception_interrupt(mocker, capsys, snapshot):
    mocked_process_run = mocker.patch(
        "subprocess.run",
        autospec=True,
        side_effect=KeyboardInterrupt("Ctrl+C"),
    )
    mocked_sys_exit = mocker.patch("sys.exit")
    args = ("bash", "-c", "echo 1")
    process.run(*args)
    mocked_process_run.assert_called_once_with(args, check=True, encoding="UTF-8")
    mocked_sys_exit.assert_called_once_with(constants.ErrorCode.KEYBOARD_INTERRUPT)
    captured = capsys.readouterr()
    snapshot.assert_match(process.clean_output(captured.out))
