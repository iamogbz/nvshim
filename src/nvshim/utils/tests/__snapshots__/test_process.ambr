# name: test_process_run_handles_exception_interrupt
  'Interrupted. Ctrl+C'
---
